### PR TITLE
Add and test DEMO mode for public use. Fix a lexer bug in new feature.

### LINKED ISSUE #9659: Compiler crash if using new literal features without enabling them
If you attempt to use the new lexical features without enabling them, the compiler attempts to produce a diagnostic without a proper fill-in, resulting in a string.Format exception.

### PR DESCRIPTION
**Pull request by gafter**
Fixes #9659 #9662 

@jaredpar, @dotnet/roslyn-compiler Please review.
@davkean This targets future-stabilization
CC @MattGertz For Ask Mode approval.

### CODE REVIEW

**Comment by MattGertz** [comment]
Approved pending CRs & tests.

File: src/Compilers/CSharp/Portable/CSharpParseOptions.cs [hunk 1]
```diff
@@ -210,10 +210,23 @@ public new CSharpParseOptions WithFeatures(IEnumerable<KeyValuePair<string, stri
...
+                {
+                    case MessageID.IDS_FeatureBinaryLiteral:
+                    case MessageID.IDS_FeatureDigitSeparator:
+                    case MessageID.IDS_FeatureLocalFunctions:
+                    case MessageID.IDS_FeatureRefLocalsReturns:
+                    case MessageID.IDS_FeaturePatternMatching:
+                        return true;
+                    case MessageID.IDS_FeaturePatternMatching2:
+                        // in "experimental" mode additionally enable experimental proposed new C# 7 language features.
+                        return PreprocessorSymbols.Contains("__DEMO_EXPERIMENTAL__");
```
Comment (cston) [inline src/Compilers/CSharp/Portable/CSharpParseOptions.cs]: Should it be sufficient to set "__DEMO_EXPERIMENTAL__" rather than both?

Comment (cston) [inline src/Compilers/CSharp/Portable/CSharpParseOptions.cs]: Should it be sufficient to set ```"__DEMO_EXPERIMENTAL__"``` rather than both?

**Comment by cston** [comment]
LGTM

File: src/Compilers/CSharp/Portable/CSharpParseOptions.cs (hunk 1 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/CSharpParseOptions.cs]: Good question. Given the time pressure, I'd rather not redesign the feature.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/CSharpParseOptions.cs]: It feel that we should ```break``` if contains returns false.

File: src/Compilers/CSharp/Portable/Parser/Lexer.cs [hunk 2]
```diff
@@ -898,12 +898,14 @@ private void CheckFeatureAvailability(MessageID feature)
                 return;
             }
 
-            if (feature.RequiredFeature() != null)
+            var requiredFeature = feature.RequiredFeature();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/Lexer.cs]: Please spell out the type.

File: src/Compilers/CSharp/Portable/Parser/Lexer.cs [hunk 3]
```diff
@@ -898,12 +898,14 @@ private void CheckFeatureAvailability(MessageID feature)
...
+            if (requiredFeature != null)
             {
                 if (!options.IsFeatureEnabled(feature))
                 {
-                    this.AddError(ErrorCode.ERR_FeatureIsExperimental, feature.Localize());
+                    this.AddError(ErrorCode.ERR_FeatureIsExperimental, feature.Localize(), requiredFeature);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/Lexer.cs]: It looks like there is one more call site with similar issue - Binder_Symbols.cs(1956)

**Comment by VSadov** [comment]
LGTM

File: src/Compilers/CSharp/Portable/Parser/Lexer.cs (hunk 3 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Parser/Lexer.cs]: I'll fix that, but I can't test it, as none of its call sites are for features that have feature flags.

**Comment by gafter** [comment]
@AlekseyTs Does this look good to you now?

**Comment by AlekseyTs** [comment]
LGTM

**Comment by gafter** [comment]
retest vsi please

**Comment by davkean** [comment]
@dotnet-bot retest prtest/win/vsi/p3 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/dotnet_roslyn_prtest_win_vsi_p3/475/
// Retest reason: Timeout cloning repo from GitHub. Retry the job.

**Pull request merged**